### PR TITLE
dojson: decorator for ignoring None value

### PR DESCRIPTION
**Pull request by jma**
In some cases it can be interesting to have conditional properties. For example, given the length of the ISBN identifier, it will be assigned to isbn10 or isbn13. We want in this case:
{"isbn10": "0123456789"} or {"isbn13": "01234567890123"}.

Signed-off-by: Johnny Mariéthoz johnny.mariethoz@rero.ch

### CODE REVIEW

**Comment by jirikuncar** [comment]
@jma can you please add specific test for this feature?

**Comment by jma** [comment]
The test has been added. I hope that is correct. 
The following command:

```
setup.py test
```

is working.

File: tests/test_core.py [hunk 1]
```diff
@@ -135,3 +135,16 @@ def test_simple_record_from_xml():
...
+    """Test none value."""
+    overdo = dojson.Overdo()
+
+    @overdo.over('024', '^024..')
+    def match(self, key, value):
+        return None
+
+    data = overdo.do({'0247_': 'this should not be added'})
+
+    assert data.get('024', "not exists") == "not exists"
```
Comment (jirikuncar) [inline tests/test_core.py]: What about being more explicit `assert '024' not in data, '024 should not be added to data'`?

File: dojson/overdo.py [hunk 2]
```diff
@@ -55,7 +55,9 @@ def do(self, blob):
 
         for key, value in iteritems(blob):
             for name, creator, field in self._query(key):
-                output[name] = creator(output, key, value)
+                val = creator(output, key, value)
```
Comment (jirikuncar) [inline dojson/overdo.py]: Another crazy idea that would allow storing `None` values would be to raise e.g. `KeyError`.

### Example

``` python
overdo = dojson.Overdo()

@overdo.over('024', '^024..')
def match(self, key, value):
    raise KeyError(key)  # or custom IgnoreKey(key)
```

``` diff
+try
-output[name] = creator(output, key, value)
+    output[name] = creator(output, key, value)
+except KeyError:  # or IgnoreKey
+    # logging or pass
```

File: tests/test_core.py (hunk 1 above)
Comment (jma) [inline tests/test_core.py]: Yes it is better I will commit soon.

File: dojson/overdo.py (hunk 2 above)
Comment (jma) [inline dojson/overdo.py]: I do not like this approach as KeyError is usually generated when we get the value. In this case is when we set the value.

An other approach can be similar as utils.filter_value. But I don't see how to implement this as a decorator. Any idea?

Comment (jirikuncar) [inline dojson/overdo.py]: and the other approach with custom `IgnoreKey` exception?

Comment (jirikuncar) [inline dojson/overdo.py]:
``` python
def ignore_value(value=None):
    ...
        if value is None and result is None or result == value:
            raise IgnoreKey(key)
    ...
    return decorator
```

Comment (jma) [inline dojson/overdo.py]: Nice approach. Do you think we have to do this in the same decorator or we have to split into two functions: ignore_value, ignore_key?

Comment (jirikuncar) [inline dojson/overdo.py]: IMHO we are ignoring key based on value. Maybe the exception should be called `IgnoredKey(key)`. WDYT?

Comment (jma) [inline dojson/overdo.py]: That sounds cool. I will try this approach. Do you know how to run a simple test:

```
py.test tests
```

does not work.

Comment (jirikuncar) [inline dojson/overdo.py]: just `py.test` should work (`py.test tests/test_<name>.py`)

``` console
$ py.test -k <name-of-tests>
```

**Comment by jma** [comment]
I modified the code to add a decorator to generate a IgnoreKey exception.

Note: when I try to run:

```
py.test tests/test_core.py
```

I got:

```
usage: py.test [options] [file_or_dir] [file_or_dir] [...]
py.test: error: unrecognized arguments: --clearcache --pep8 --cov=dojson --cov-report=term-missing py
```

**Comment by jirikuncar** [comment]
@jma you need to have `pytest-pep8`, `pytest-cache`, `pytest-cov` installed.

**Comment by jma** [comment]
@jirikuncar all of these are installed. I think, I have a problem as I got the same message with:

```
py.test --version
```

**Comment by jma** [comment]
@jirikuncar removing the pytest.ini file make the command working.

**Comment by jirikuncar** [comment]
@jma please have a look what dependencies are needed in `setup.py` for tests. If you are running `python setup.py test` they are installed as eggs in current directory, but if you want to run `py.test` you need to install them in your virtualenv.

**Comment by jma** [comment]
@jirikuncar done, it works thanks.

File: dojson/utils.py [hunk 3]
```diff
@@ -13,6 +13,25 @@
...
+
+
+def ignore_value(f):
+    """Remove key for None value."""
+    @functools.wraps(f)
+    def wrapper(self, key, value, **kwargs):
+        result = f(self, key, value, **kwargs)
+        if result is None:
+            raise IgnoreKey(key)
+        else:
```
Comment (jirikuncar) [inline dojson/utils.py]: IMHO you don't need the `else` branch here (improves readability).

File: dojson/utils.py [hunk 4]
```diff
@@ -13,6 +13,25 @@
...
 
 
+class IgnoreKey(Exception):
+
+    """The corresponding key has been ignored."""
+
+    pass
+
+
+def ignore_value(f):
```
Comment (jirikuncar) [inline dojson/utils.py:27]: @jma shall we accept optional argument with value that should be ignored otherwise we can rename it to `ignore_none_value`.

Comment (jirikuncar) [inline dojson/utils.py:27]: I see that I haven't add this configuration in my initial design of `filter_values`. Hence I think it can stay. Thanks!

Comment (jma) [inline dojson/utils.py:27]: @jirikuncar I removed the useless "else".

Comment (jma) [inline dojson/utils.py:27]: Thus can we keep the original name `ignore_value`? In this case is the PR ready for integration?

Comment (jirikuncar) [inline dojson/utils.py:27]: @jma yes, just few cosmetics. Can you please add `.. versionadded:: 0.2.0` (see http://sphinx-doc.org/markup/para.html#directive-versionadded)

File: tests/test_core.py [hunk 5]
```diff
@@ -135,3 +136,17 @@ def test_simple_record_from_xml():
...
+def test_none_value():
+    """Test none value."""
+    overdo = dojson.Overdo()
+    from dojson.utils import ignore_value
+
+    @overdo.over('024', '^024..')
+    @ignore_value
+    def match(self, key, value):
+        return None
+
```
Comment (jirikuncar) [inline tests/test_core.py:150]: can you also add a test when the decorator is used but the return value is not None?

**Comment by jirikuncar** [comment]
@jma please amend commit message so we can auto-generate RELEASE-NOTES. Thanks

```
utils: decorator for ignoring None value

* NEW Adds new decorator `ignore_value` that ... # add more description here

Reviewed-by: Jiri Kuncar <jiri.kuncar@cern.ch>
Signed-off-by: you
```

File: tests/test_core.py [hunk 6]
```diff
@@ -135,3 +136,31 @@ def test_simple_record_from_xml():
...
+    overdo = dojson.Overdo()
+    from dojson.utils import ignore_value
+
+    @overdo.over('024', '^024..')
+    @ignore_value
+    def match(self, key, value):
+        return value
+
+    data = overdo.do({'0247_': 'valid value'})
+    assert data.get("024") == "valid value"
```
Comment (jirikuncar) [inline tests/test_core.py]: @jma I'm sorry that I'm so picky, but can you use single quotes here for consistency?
